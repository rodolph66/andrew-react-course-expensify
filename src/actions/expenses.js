import uuid from 'uuid'
import database from '../firebase/firebase'
// EXPENSE ACTIONS ######################################

// ADD_EXPENSE ********************************************
export const addExpense = (expense) => ({
  type: 'ADD_EXPENSE',
  expense
})

export const startAddExpense = (expenseData = {}) => {
  return (dispatch) => {
    const {description='', note='', amount=0, createdAt=0} = expenseData
    const expense = { description, note, amount, createdAt }
    // the "return" is optional, here added to be used in the test files
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({id: ref.key, ...expense}))
    })
  }
}

// ********************************************************

// REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
  type: 'REMOVE_EXPENSE',
  id
})

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }))
    })
  }
}

// ********************************************************

// EDIT_EXPENSE
export const editExpense = (id , updates) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
})

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update(updates).then(() => {
      dispatch(editExpense(id, updates))
    })
  }
}

// ********************************************************

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
})

export const startSetExpenses = () => {
  return (dispatch) => {
    // the "return" is optional, here added to be used in the test files
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = []
      snapshot.forEach(element => {
        expenses.push({
          id: element.key,
          ...element.val()
        })
      });
      dispatch(setExpenses(expenses))
    })
  }
}

// ********************************************************

/**
 * activities to update the store involves:
 * 1. component calls action generator
 * 2. action generator returns a function (to return func middleware is needed)
 * 3. component dispatches the function
 * 4. the function performs a task of adding/updating/deleting and expesne
 * 5. on its completiong the funciton dispatch an action to update store
 */