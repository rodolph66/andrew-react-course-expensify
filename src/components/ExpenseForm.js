import React from 'react'
import moment from 'moment'
import { SingleDatePicker } from 'react-dates'

export default class ExpenseForm extends React.Component {
  // note 2
  state = {
    description: this.props.expense ? this.props.expense.description : '',
    note: this.props.expense ? this.props.expense.note : '',
    amount: this.props.expense ? (this.props.expense.amount/100).toString() : '',
    createdAt: this.props.expense ? moment(this.props.expense.createdAt) : moment(),
    calendarFocused: false,
    error: '',
    buttonLabel: this.props.expense ? 'Save Changes' : 'Add Expense'
  }

  onDescriptionChange = (e) => {
    // note 1
    const description = e.target.value
    this.setState(() => ({description}))
  }

  onNoteChange = (e) => {
    //note 1
    const note = e.target.value
    this.setState(() => ({note}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({amount}))
    }
  }

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }))
    }
  }

  onFocusedChange = ({ focused }) => {
    this.setState({ calendarFocused: focused })
  }

  onSubmit = (e) => {
    e.preventDefault()
    if (!this.state.description || !this.state.amount) {
      this.setState({error: 'Please provide description and amount'})
    } else {
      this.setState({error: ''})
      this.props.onSubmit({
        description: this.state.description,
        amount: parseInt(this.state.amount * 100, 10),
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      })
    }
  }

  render() {
    return (  
      <div>
        {this.state.error && <h4>{this.state.error}</h4>}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange} 
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusedChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
          >
          </textarea>
          <button>{this.state.buttonLabel}</button>
        </form>
      </div>
    )
  }
}



// note 1: the target value must be saved in a var/const before setState

// note 2: if anything fails try the following which is what is used in the course
/*
  constructor(props) {
    super(props)
    this.state = {
      description: props.expense ? props.expense.description : ''
      note: props.expense ? props.expense.note : '',
      amount: props.expense ? (props.expense.amount/100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    }
  }
*/