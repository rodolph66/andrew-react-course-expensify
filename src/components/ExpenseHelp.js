import React from 'react'

const ExpenseHelp = () => (
  <div>
    This is from my HELP
  </div>
)

export default ExpenseHelp