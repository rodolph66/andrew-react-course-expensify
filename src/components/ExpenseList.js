import React from 'react'
import { connect } from 'react-redux'
import ExpenseListItem from './ExpenseListItem'
import selectExpenses from '../selectors/expenses'

// Note: "export" is required  so that the component can be used in testing (it is optional)
export const ExpenseList = (props) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-for-mobile">Expenses</div>
      <div className="show-for-desktop">Expense</div>
      <div className="show-for-desktop">Amount</div>
    </div>
    <div className="list-body">
      { props.expenses.length === 0 ? (
        <div className="list-item list-item--message">
          <span>No expenses</span>
        </div>
      ) : (
        props.expenses.map( expense => ( <ExpenseListItem key={expense.id} {...expense} /> ))
      )}
    </div>
  </div>
)


// OPTION 3 with the selectExpenses used to filter state values
const mapStateToProps = state => ({expenses: selectExpenses(state.expenses, state.filters)})

export default connect(mapStateToProps)(ExpenseList)


/*
// OPTION 3
const mapStateToProps = state => ({expenses: state.expenses, filters: state.filters})
/*

/* OPTION 2
export default connect((state) => ({expenses: state.expenses}))(ExpenseList)
*/

/* OPTION 1
// connect(param)(component) allows the component to access part of the store. 
// This part is specified in the param as is passed to the component as props
const ConnectedExpenseList = connect((state) => ({expenses: state.expenses}))(ExpenseList)

export default ConnectedExpenseList
*/