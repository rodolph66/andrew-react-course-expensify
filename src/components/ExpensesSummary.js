import React from 'react'
import numeral from 'numeral'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import getExpensesTotal from '../selectors/expense-total'
import selectExpenses from '../selectors/expenses'


// Note: "export" is required  so that the component can be used in testing (it is optional)
export const ExpensesSummary = ({expenseCount, expensesTotal}) => {
  const expenseWord = expenseCount===1 ? 'expense' : 'expenses'
  const formatedTotal = numeral(expensesTotal/100).format('$0,0.00')
  return (
    <div className="page-header">
      <div className="content-container">
        {expenseCount>0 ? (
          <h1 className="page-header__title">
            Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{formatedTotal}</span>
          </h1>
        ) : (
          <h1 className="page-header__title">There are no expenses to view</h1>
        )}
        <div className="page-header__action">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
    </div>
  )
}

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters)
  return ({
    expenseCount: visibleExpenses.length,
    expensesTotal: getExpensesTotal(visibleExpenses)
  })
}

export default connect(mapStateToProps)(ExpensesSummary)



/* Alternate version of the JSX
<div>
  {(() => {
    switch(expenseCount) {
      case 0 : 
        return <p>There are no expenses to view</p>
        break
      case 1 : 
        return (
          <p> Viewing 1 exepense totalling {expensesTotal}</p>
        )
        break
      default: 
      return (
        <p> Viewing {expenseCount} exepense totalling {expensesTotal}</p>
      )
    }
  })()}
</div>
*/