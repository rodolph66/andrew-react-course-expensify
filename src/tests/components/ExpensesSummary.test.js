import React from 'react'
import { shallow } from 'enzyme'
import { ExpensesSummary } from '../../components/ExpensesSummary'
import expenses from '../fixtures/expenses'


test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={1} expensesTotal={235} />)
  expect(wrapper).toMatchSnapshot()
})

test('should render ExpenseList with empty message', () => {
  const wrapper = shallow(<ExpensesSummary expensesCount={23} expensesTotal={150050099} />)
  expect(wrapper).toMatchSnapshot()
})