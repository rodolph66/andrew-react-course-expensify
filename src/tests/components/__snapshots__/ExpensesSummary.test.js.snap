// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with empty message 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      There are no expenses to view
    </h1>
    <div
      className="page-header__action"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseList with expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      There are no expenses to view
    </h1>
    <div
      className="page-header__action"
    >
      <Link
        className="button"
        replace={false}
        to="/create"
      >
        Add Expense
      </Link>
    </div>
  </div>
</div>
`;
